### DISCUSSION TITLE
Two more reserved attributes for easier error handling

### DISCUSSION THREAD

**Discussion by thargol1** [dotnet/csharplang#4623] [Language Ideas]
# Two more reserved attributes for easier error handling

A common pattern is to use Throw helpers for generating consistent error checks and error messages.

An example:
``` csharp
public static class Throw
{
	public static void WhenIsNullOrEmpty(string value, string paramName)
	{
		if (string.IsNullOrEmpty(value))
		{
			throw new ArgumentException($"'{paramName}' cannot be null or empty.", paramName);
		}
	}
}
```
And is called like this:
``` csharp
void MyDemo(string? name)
{
	Throw.WhenIsNullOrEmpty(name, nameof(name));
}
```

This example has two problems:
1. The null checks in the compiler don't see that this function does not return when value is null or empty.
2. It's a lot of typing for little functionality.

Suggestion 1: Add a new attribute to indicate that a function does not return if an argument is null or empty:
```csharp
public static void WhenIsNullOrEmpty([DoesNotReturnWhenNullOrEmpty] string value, string paramName)
```
or
```csharp
public static void WhenIsNull([DoesNotReturnWhenNull] object value, string paramName)
```

Suggestion 2: Add a new attribute that gets the source code from a parameter like `CallerMemberNameAttribute` returns the calling function name. The `(0)` indicates which parameter to get.
```csharp
public static void WhenIsNullOrEmpty(string value, [ParameterSource(0)] string paramName = "")
```
So
```csharp
Throw.WhenIsNullOrEmpty(name);
```
Lowers to:
```csharp
Throw.WhenIsNullOrEmpty(name, "name");
```
Or
```csharp
Throw.WhenIsNullOrEmpty(options.Name);
```
Lowers to:
```csharp
Throw.WhenIsNullOrEmpty(options.Name, "options.Name");
```

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/issues/287

**Comment by jnm2** [comment]
> The null checks in the compiler don't see that this function does not return when value is null or empty.

There's already an annotation that achieves the null side of this. Put `[NotNull]` on the parameter; it's a postcondition that applies if the function returns. It means exactly what you want `[DoesNotReturnWhenNull]` to mean.

**Reply by 333fred** [reply]
The combination of this and HaloFour's answer is exactly what you're looking for @thargol1 (with C# 10 support for `CallerExpressionAttribute`):

```cs
public static void ThrowWhenNull([NotNull] object value, [CallerExpression("value")] string valueExpression = "")
{
    if (value is null) throw new ArgumentNullException(valueExpression);
}
```

**Comment by thargol1** [comment]
Thanks for the answers.

Perhaps https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullattribute?view=net-5.0 should be expanded with an example or better clarification.

It's exactly what I asked for, but I didn't understand that from the documentation.

(This issue can be closed.)

**Reply by 333fred** [reply]
@BillWagner making sure you see this feedback.

**Reply by BillWagner** [reply]
Thanks.  [This article](https://docs.microsoft.com/dotnet/csharp/language-reference/attributes/nullable-analysis#specify-preconditions-allownull-and-disallownull) explains the use of all these attributes. (It hasn't been updated for C# 10 though).

I like the sample @thargol1 showed above. I'll add that (unless @thargol1 wants to submit a PR as this was his idea).

I'll work with our team to ensure that the pages on each of these attributes links to the conceptual doc that explains how to use all the attributes.

**Reply by thargol1** [reply]
Go ahead Bill and add it. I don't need kudos.
